common/variant/src/main/java/org/apache/spark/types/variant/VariantShreddingWriter.java

```diff
@@ -0,0 +1,298 @@
+  // to populate it.
+  public interface ShreddedResult {
+    // Create an array. The elements are the result of shredding each element.
+    void addArray(VariantSchema schema, List<ShreddedResult> array);
+    // Create an object. The values are the result of shredding each field, order by the index in
+    // objectSchema. Missing fields are populated with an empty result.
+    void addObject(VariantSchema schema, ShreddedResult[] values);
+    void addVariantValue(VariantSchema schema, byte[] result);
+    // Add a scalar to typed_value. The type of Object depends on the scalarSchema in the shredding
+    // schema.
+    void addScalar(VariantSchema schema, Object result);
+    void addMetadata(VariantSchema schema, byte[] result);
+  }
+
+  public interface ShreddedResultBuilder {
+    ShreddedResult createEmpty(VariantSchema schema);
+
+    // If true, we will shred decimals to a different scale or to integers, as long as they are
+    // numerically equivalent. Similarly, integers will be allowed to shred to decimals.
+    boolean allowNumericScaleChanges();
+  }
+
+  /**
+   * Converts an input variant into shredded components. Returns the shredded result, as well
+   * as the original Variant with shredded fields removed.
+   * `dataType` must be a valid shredding schema, as described in common/variant/shredding.md.
+   */
+  public static ShreddedResult castShredded(
+          Variant v,
+          VariantSchema schema,
+          ShreddedResultBuilder builder) {
+    VariantUtil.Type variantType = v.getType();
+    ShreddedResult result = builder.createEmpty(schema);
+
+    if (schema.topLevelMetadataIdx >= 0) {
+      result.addMetadata(schema, v.getMetadata());
+    }
+
+    if (schema.arraySchema != null && variantType == VariantUtil.Type.ARRAY) {
+      // The array element is always a struct containing untyped and typed fields.
+      VariantSchema elementSchema = schema.arraySchema;
+      int size = v.arraySize();
+      ArrayList<ShreddedResult> array = new ArrayList<>(size);
+      for (int i = 0; i < size; ++i) {
+        ShreddedResult shreddedArray = castShredded(v.getElementAtIndex(i), elementSchema, builder);
+        array.add(shreddedArray);
+      }
+      result.addArray(schema, array);
+    } else if (schema.objectSchema != null && variantType == VariantUtil.Type.OBJECT) {
+      Map<String, VariantSchema.ObjectField> objectSchema = schema.objectSchema;
+      int maxSize = Math.min(v.objectSize(), objectSchema.size());
+      ShreddedResult[] shreddedValues = new ShreddedResult[objectSchema.size()];
+
+      // Create a variantBuilder for any mismatched fields.
+      VariantBuilder variantBuilder = new VariantBuilder(false);
+      ArrayList<VariantBuilder.FieldEntry> fieldEntries = new ArrayList<>();
+      // Keep track of which schema fields we actually found in the Variant value.
+      Set<String> presentKeys = new HashSet<String>();
```

Comment:
Okay, I was on the fence about how much of an effort to put into efficiency here.

I modified the interface so that the shredding schema constructor takes an array of name-schema pairs. I like this better than the map in any case, since my code was assuming that the indices in the map were dense and corresponded to the array being returned in the shredded result, and this makes that much clearer. I now construct the name->field mapping in the constructor. Let me know what you think.

I didn't bother with trying to create an ordered list of fields, I just iterate over the array a second time to fill in nulls if not all fields matched. I think it could be a bit more efficient to sort the list and step through the list and the Variant object together, but it adds more complexity and branching to the loop, and I'm not sure that it would be significantly faster than the simple second iteration over the loop to fill in null values. Let me know if you had something different in mind.